### PR TITLE
[fix](merge-on-write) fix that be may coredump when sequence column is null

### LINKED ISSUE #16833: [Bug] (merge-on-write)be may coredump when sequence column is null
### Search before asking

- [X] I had searched in the [issues](https://github.com/apache/doris/issues?q=is%3Aissue) and found no similar issues.

### Version

master

### What's Wrong?

Query id: 0-0 ***
Aborted at 1675069046 (unix time) try "date -d @1675069046" if you are using GNU date ***
Current BE git commitID: aaa19e9b4d ***
SIGABRT unkown detail explain (@0x1cd9aa) received by PID 1890730 (TID 0x7f68d2a1f700) from PID 1890730; stack trace: ***
 0# doris::signal::(anonymous namespace)::FailureSignalHandler(int, siginfo_t*, void*) at /root/code/doris/be/src/common/signal_handler.h:420
 1# 0x00007F6AC60A3400 in /lib64/libc.so.6
 2# gsignal in /lib64/libc.so.6
 3# abort in /lib64/libc.so.6
 4# 0x000055E5F9D54CD9 in /root/code/doris/output/be/lib/doris_be
 5# 0x000055E5F9D4A2ED in /root/code/doris/output/be/lib/doris_be
 6# google::LogMessage::SendToLog() in /root/code/doris/output/be/lib/doris_be
 7# google::LogMessage::Flush() in /root/code/doris/output/be/lib/doris_be
 8# google::LogMessageFatal::~LogMessageFatal() in /root/code/doris/output/be/lib/doris_be
 9# doris::MemPool::allocate(long, bool) at /root/code/doris/be/src/runtime/mem_pool.h:108
10# doris::FieldTypeTraits<(doris::FieldType)13>::deep_copy(void*, void const*, doris::MemPool*) at /root/code/doris/be/src/olap/types.h:1426
11# doris::ScalarTypeInfo::deep_copy(void*, void const*, doris::MemPool*) const at /root/code/doris/be/src/olap/types.h:109
12# doris::segment_v2::(anonymous namespace)::BloomFilterIndexWriterImpl<(doris::FieldType)17>::add_values(void const*, unsigned long) at /root/code/doris/be/src/olap/rowset/segment_v2/bloom_filter_index_writer.cpp:84
13# doris::PrimaryKeyIndexBuilder::add_item(doris::Slice const&) at /root/code/doris/be/src/olap/primary_key_index.cpp:47
14# doris::segment_v2::SegmentWriter::append_block(doris::vectorized::Block const*, unsigned long, unsigned long) at /root/code/doris/be/src/olap/rowset/segment_v2/segment_writer.cpp:188
15# doris::BetaRowsetWriter::_add_block(doris::vectorized::Block const*, std::unique_ptr<doris::segment_v2::SegmentWriter, std::default_delete<doris::segment_v2::SegmentWriter> >*) at /root/code/doris/be/src/olap/rowset/beta_rowset_writer.cpp:574
16# doris::BetaRowsetWriter::flush_single_memtable(doris::vectorized::Block const*) at /root/code/doris/be/src/olap/rowset/beta_rowset_writer.cpp:699
17# doris::MemTable::_do_flush(long&) at /root/code/doris/be/src/olap/memtable.cpp:488
18# doris::MemTable::flush() at /root/code/doris/be/src/olap/memtable.cpp:455
19# doris::FlushToken::_flush_memtable(doris::MemTable*, long) at /root/code/doris/be/src/olap/memtable_flush_executor.cpp:90
20# doris::MemtableFlushTask::run() at /root/code/doris/be/src/olap/memtable_flush_executor.cpp:40
21# doris::ThreadPool::dispatch_thread() at /root/code/doris/be/src/util/threadpool.cpp:534
22# void std::_invoke_impl<void, void (doris::ThreadPool::&)(), doris::ThreadPool&>(std::_invoke_memfun_deref, void (doris::ThreadPool::&)(), doris::ThreadPool&) at /var/local/ldb-toolchain/bin/../lib/gcc/x86_64-linux-gnu/11/../../../../include/c++/11/bits/invoke.h:74
23# std::_invoke_result<void (doris::ThreadPool::&)(), doris::ThreadPool&>::type std::_invoke<void (doris::ThreadPool::&)(), doris::ThreadPool&>(void (doris::ThreadPool::&)(), doris::ThreadPool&) at /var/local/ldb-toolchain/bin/../lib/gcc/x86_64-linux-gnu/11/../../../../include/c++/11/bits/invoke.h:96
24# void std::Bind<void (doris::ThreadPool::(doris::ThreadPool))()>::_call<void, , 0ul>(std::tuple<>&&, std::_Index_tuple<0ul>) at /var/local/ldb-toolchain/bin/../lib/gcc/x86_64-linux-gnu/11/../../../../include/c++/11/functional:420
25# void std::_Bind<void (doris::ThreadPool::(doris::ThreadPool))()>::operator()<, void>() at /var/local/ldb-toolchain/bin/../lib/gcc/x86_64-linux-gnu/11/../../../../include/c++/11/functional:503
26# void std::_invoke_impl<void, std::_Bind<void (doris::ThreadPool::(doris::ThreadPool))()>&>(std::_invoke_other, std::_Bind<void (doris::ThreadPool::(doris::ThreadPool))()>&) at /var/local/ldb-toolchain/bin/../lib/gcc/x86_64-linux-gnu/11/../../../../include/c++/11/bits/invoke.h:61
27# std::enable_if<is_invocable_r_v<void, std::Bind<void (doris::ThreadPool::(doris::ThreadPool))()>&>, void>::type std::_invoke_r<void, std::_Bind<void (doris::ThreadPool::(doris::ThreadPool))()>&>(std::_Bind<void (doris::ThreadPool::(doris::ThreadPool))()>&) at /var/local/ldb-toolchain/bin/../lib/gcc/x86_64-linux-gnu/11/../../../../include/c++/11/bits/invoke.h:117
28# std::_Function_handler<void (), std::_Bind<void (doris::ThreadPool::(doris::ThreadPool))()> >::_M_invoke(std::_Any_data const&) at /var/local/ldb-toolchain/bin/../lib/gcc/x86_64-linux-gnu/11/../../../../include/c++/11/bits/std_function.h:291
29# std::function<void ()>::operator()() const at /var/local/ldb-toolchain/bin/../lib/gcc/x86_64-linux-gnu/11/../../../../include/c++/11/bits/std_function.h:560
30# doris::Thread::supervise_thread(void*) at /root/code/doris/be/src/util/thread.cpp:454
31# start_thread in /lib64/libpthread.so.0

### What You Expected?

fix it

### How to Reproduce?

load data into unique with merge-on-write table when sequence column is null.

### Anything Else?

_No response_

### Are you willing to submit PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### PR DESCRIPTION
**Pull request by liaoxin01**
# Proposed changes

Issue Number: close #16833 

## Problem summary

To facilitate the use of the primary key index, encode the seq column to the minimum value of the corresponding length when the seq column is null.

## Checklist(Required)

* [ ] Does it affect the original behavior
* [ ] Has unit tests been added
* [ ] Has document been added or modified
* [ ] Does it need to update dependencies
* [ ] Is this PR support rollback (If NO, please explain WHY)

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by morningman** [comment]
Please add test case

**Comment by liaoxin01** [comment]
> Please add test case

Test case has been added.

File: be/src/olap/rowset/segment_v2/segment_writer.cpp [hunk 1]
```diff
@@ -389,6 +388,22 @@ std::string SegmentWriter::_full_encode_keys(
...
+    // To facilitate the use of the primary key index, encode the seq column
+    // to the minimum value of the corresponding length when the seq column
+    // is null
+    if (UNLIKELY(!field)) {
+        encoded_keys->push_back(KEY_NULL_FIRST_MARKER);
+        size_t seq_col_length = _tablet_schema->column(_tablet_schema->sequence_col_idx()).length();
+        encoded_keys->append(seq_col_length, KEY_MINIMAL_MARKER);
+        return;
+    }
+    encoded_keys->push_back(KEY_NORMAL_MARKER);
```
Comment (carlvinhust2012) [inline be/src/olap/rowset/segment_v2/segment_writer.cpp:403]: No need to check the validity of the pointers here?

Comment (liaoxin01) [reply be/src/olap/rowset/segment_v2/segment_writer.cpp:403]: check encoded_keys? encoded_keys is an output parameter, the caller will ensure the validity.

File: be/src/olap/rowset/segment_v2/segment_writer.cpp [hunk 2]
```diff
@@ -389,6 +388,22 @@ std::string SegmentWriter::_full_encode_keys(
...
+void SegmentWriter::_encode_seq_column(const vectorized::IOlapColumnDataAccessor* seq_column,
+                                       size_t pos, string* encoded_keys) {
+    auto field = seq_column->get_data_at(pos);
+    // To facilitate the use of the primary key index, encode the seq column
+    // to the minimum value of the corresponding length when the seq column
+    // is null
+    if (UNLIKELY(!field)) {
+        encoded_keys->push_back(KEY_NULL_FIRST_MARKER);
+        size_t seq_col_length = _tablet_schema->column(_tablet_schema->sequence_col_idx()).length();
+        encoded_keys->append(seq_col_length, KEY_MINIMAL_MARKER);
```
Comment (zhannngchen) [inline be/src/olap/rowset/segment_v2/segment_writer.cpp:400]: if the sequence columns is type of INT, is it work well if user set negative values?

Comment (liaoxin01) [reply be/src/olap/rowset/segment_v2/segment_writer.cpp:400]: There is no problem.  Because the signed number will reverse the sign bit once.

**Review by zhannngchen — APPROVED** [review]
LGTM

**Pull request merged**